File: backend/app/models/channel.py
# backend/app/models/channel.py

from sqlalchemy import String, DateTime, Integer, Boolean, UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
from typing import Optional

from app.extensions import db
from app.models.base import Base
from app.models.account import StatsTrackAccountGuid
from app.models.stats import StatsAggregatedChannel, StatsTrackEventChannel


class Channel(Base):
    __tablename__ = "channel"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    id_text: Mapped[str] = mapped_column(String(15), unique=True, nullable=False)
    slug: Mapped[Optional[str]] = mapped_column(String(100), unique=True)
    feed_id: Mapped[int] = mapped_column(Integer, db.ForeignKey("feed.id"), nullable=False)
    podcast_index_id: Mapped[int] = mapped_column(Integer, nullable=False)
    podcast_guid: Mapped[Optional[str]] = mapped_column(UUID(as_uuid=False), unique=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False, unique=True)
    sortable_title: Mapped[Optional[str]] = mapped_column(String(255), unique=True)
    medium_id: Mapped[Optional[int]] = mapped_column(Integer, db.ForeignKey("medium.id"), nullable=True)
    has_podcast_index_value: Mapped[bool] = mapped_column(Boolean, default=False)
    has_value_time_splits: Mapped[bool] = mapped_column(Boolean, default=False)

    feed = relationship("Feed", back_populates="channels")
    medium = relationship("Medium", back_populates="channels")
    categories = relationship(
        "ChannelCategory",
        back_populates="channel",
        cascade="all, delete-orphan"
    )
    items = relationship(
        "Item",
        back_populates="channel"
    )
    stats = relationship(
        "StatsAggregatedChannel",
        back_populates="channel",
        cascade="all, delete-orphan"
    )
    events = relationship(
        "StatsTrackEventChannel",
        back_populates="channel",
        cascade="all, delete-orphan"
    )


class ChannelCategory(Base):
    __tablename__ = "channel_category"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    channel_id: Mapped[int] = mapped_column(Integer, db.ForeignKey("channel.id", ondelete="CASCADE"), nullable=False)
    category_id: Mapped[int] = mapped_column(Integer, db.ForeignKey("category.id", ondelete="CASCADE"), nullable=False)
    channel = relationship("Channel", back_populates="categories")
    category = relationship("Category", back_populates="channels")

File: frontend/types/channel.ts
import { Feed } from './feed';
import {Item} from './item';
import {Stats} from './stats';
export interface Channel {
    id: number;
    id_text: string;
    slug: string;
    feed_id: number;
    podcast_index_id?: number;
    podcast_guid: string;
    title: string;
    sortable_title: string;
    medium_id: number;
    has_podcast_idex_value: boolean;
    has_value_time_splits: boolean
    category: Category[];
    medium: Medium;
}

export interface Category {
    id: number;
    display_name: string;
    mapping_key: string;
    parent_id?: number;
    slug: string;
}

export interface Medium {
    id: number;
    value: string;
}

export interface ChannelData {
  id: number;
  title: string;
  podcast_index_id: number;
  id_text: string;
  slug: string;
  medium_id: number;
  categories: Category[];
  medium: Medium;
  feed: Feed;
  items: Item[];
  stats: Stats[];
}